### PR TITLE
Add fio2_percent and oxygen_flow_rate

### PR DESCRIPTION
**Pull request by pipliggins**

### CODE REVIEW

**Review by ekamau — CHANGES_REQUESTED** [review]
thanks @pipliggins - lots of work!
all look very good; and just a few minor changes to look at ..

File: isaric/parsers/idor-brazil.toml
Comment (pipliggins) [reply isaric/parsers/idor-brazil.toml:1]: `works_lab` is a super type, which includes people who work in an (unspecified) laboratory as well as a microbiology lab. There are therefore scenarios like in idol-brazil where both `works_lab` and `works_microbiology_lab` are present, and others where only `works_lab` is present as a _microbiology_ lab has not been specified in the data dictionary.

File: isaric/parsers/sprintsari-australia.toml
Comment (pipliggins) [reply isaric/parsers/sprintsari-australia.toml:115]: healthwork_eterm = 4 relates to a microbiology lab worker, and therefore should flag both `works_lab` and `works_microbiology_lab` as True (see below)

File: isaric/parsers/ccp-guinea.toml
Comment (pipliggins) [reply isaric/parsers/ccp-guinea.toml]: thanks for catching that, the overall_oxygen stuff should have been removed.

**Comment by pipliggins** [comment]
@sadiekelly would you mind checking the guinea parser particularly? It seems like it was merged before your review changes were applied, so it's worth going back and double checking I caught everything.

File: isaric/parsers/isaric-ecmo.toml
Comment (abhidg) [inline isaric/parsers/isaric-ecmo.toml:1128]: what is the difference between fio2 and fio2b?

Comment (pipliggins) [reply isaric/parsers/isaric-ecmo.toml:1128]: fio2b contains values as a percentage, fio2 as decimals (or possible vv.) - but both appear to contain a mix of decimals and percentage values in practice

File: isaric/parsers/ccp-cameroon.toml
Comment (pipliggins) [reply isaric/parsers/ccp-cameroon.toml:1062]: I think we want it to match m3/m6/m9, looking at the data

Comment (abhidg) [reply isaric/parsers/ccp-cameroon.toml:1062]: ^ then the regex can be 'm\d_arm_1'

Comment (pipliggins) [reply isaric/parsers/ccp-cameroon.toml]: Thanks @abhidg, I've updated this

File: isaric/parsers/ccp-drc.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-drc.toml:303]: @pipliggins I think for creating the works_lab supertype we'd only map the Y responses if the question specifically asks about working in a microbiology lab, as someone could respond 'N' to not working in a microbiology lab but they are still a lab worker, what do you think?

File: isaric/parsers/ccp-ghana.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-ghana.toml:1343]: @pipliggins just to check, the function "Percentage" converts the fractional values (0.21-1.0) to percentage (21-100%)?

File: isaric/parsers/ccp-senegal.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-senegal.toml:250]: @pipliggins I'm thinking perhaps this variable should only map to works_lab, as it is not specifically asking about micro lab work

File: isaric/parsers/idor-brazil.toml
Comment (sadiekelly) [inline isaric/parsers/idor-brazil.toml:274]: same comment for mapping only Y to works_lab where the question asks specifically about microbiology lab work

File: isaric/parsers/isaric-ccpuk.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-ccpuk.toml:86]: same comment mapping Y responses only to works_lab from the microbiology question

File: isaric/parsers/isaric-ccpuk.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-ccpuk.toml:1032]: daily_fio2b_lborres is already provided as a percentage, so is the "Percentage" function not needed here?
daily_fio2_lborres is provided as a fraction of 1 so the percentage function looks ok for this one

File: isaric/parsers/isaric-ecmo.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-ecmo.toml:281]: same comment for mapping Y responses from microbiology question

File: isaric/parsers/isaric-ecmo.toml
Comment (sadiekelly) [reply isaric/parsers/isaric-ecmo.toml:1128]: can we create a function to convert to percentage only if the value is between 0.21 and 1.0? and leave as is if between 21 and 100 (%). There are probably values in between 1 and 21 that will have to remain as reported. Would that be possible, as I don't think there's ambiguity in the units if the values are within those clearly defined ranges

File: isaric/parsers/isaric-rapid.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-rapid.toml:1556]: same comment for applying "Percentage" function here. 
Fi02 can also be captured in L/min, and an approximate conversion can be made from this to % Fi02, but it is dependent on the method of oxygen therapy. Should we create an fio2_Lmin attribute to capture this, then users can decide for themselves how/if they wish to convert the values?

File: isaric/parsers/ncov-malaysia.toml
Comment (sadiekelly) [inline isaric/parsers/ncov-malaysia.toml:132]: same comment for mapping to works_lab from microbiology question

File: isaric/parsers/ncov-malaysia.toml
Comment (sadiekelly) [inline isaric/parsers/ncov-malaysia.toml:910]: there's also Fio2 data in the variable fio_prone, daily_assess_date can be used

File: isaric/parsers/ccp-drc.toml
Comment (pipliggins) [reply isaric/parsers/ccp-drc.toml:303]: @sadiekelly so you mean if a question asks 'are you a microbiology worker?' we'd only map the yes responses to works_lab, but both Y and N for works_microbiology_lab? Yes, I think that makes sense. We might potentially need to consider this logic for other super types as well...

File: isaric/parsers/ccp-ghana.toml
Comment (pipliggins) [reply isaric/parsers/ccp-ghana.toml:1343]: Yes! It will also leave any existing percentages alone, as many datasets seem to have a mix of both regardless of the supposed unit.

File: isaric/parsers/ccp-drc.toml
Comment (sadiekelly) [reply isaric/parsers/ccp-drc.toml:303]: yes, that's right, I think that's the best way. Which of the other super types are you thinking of?

File: isaric/parsers/ccp-guinea.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-guinea.toml:595]: @pipliggins I'd noted from my initial review to use daily_corticost_cmtrt___1 = 1 to map to the visit attribute visit.treatment_dexamethasone. Maybe this now comes under this corticosteroid_type attribute? If so, daily_corticost_cmtrt___1 (dexa), ___2 (hydrocortisone), ___3 (prednisolone), and ___4 (methylprednisolone) can be mapped here also

File: isaric/parsers/ccp-guinea.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-guinea.toml:662]: to add other oxygen therapy variables to this attribute (non-inv, invasive etc.), perhaps I'm not seeing those in this amendment

**Comment by sadiekelly** [comment]
> @sadiekelly would you mind checking the guinea parser particularly? It seems like it was merged before your review changes were applied, so it's worth going back and double checking I caught everything.

@pipliggins thank you! looks ok - I just added a comment regarding the oxygen_therapy supertype to include all variables that indicate any oxygen therapy was given. I noted from my review the addition of mapping palpitations and edema but these attributes don't seem to be present in the version of the parser I'm looking at - are these attributes still being mapped?

File: isaric/parsers/ccp-drc.toml
Comment (pipliggins) [reply isaric/parsers/ccp-drc.toml:303]: I haven't thought it through... but possibly for things like oxygen therapy - @abhidg remind me, if a 'false' response is given within combinedType = any, will that override a 'True' response from earlier?

**Comment by pipliggins** [comment]
> > @sadiekelly would you mind checking the guinea parser particularly? It seems like it was merged before your review changes were applied, so it's worth going back and double checking I caught everything.
> 
> @pipliggins thank you! looks ok - I just added a comment regarding the oxygen_therapy supertype to include all variables that indicate any oxygen therapy was given. I noted from my review the addition of mapping palpitations and edema but these attributes don't seem to be present in the version of the parser I'm looking at - are these attributes still being mapped?

Thanks Sadie - these two attributes certainly aren't currently being mapped, I'm not actually sure they ever were? I've added those Guinea changes here, it's bit confusing with changes being made to the same parts of the same parser on two different branches!

**Comment by pipliggins** [comment]
Leaving this open until the difference between fio2 and o2_flow is resolved (should O2_flow be mapped as fio2_Lmin?) and we've decided whether fio2 should be included in the oxygen_therapy super type.

**Comment by sadiekelly** [comment]
> > > @sadiekelly would you mind checking the guinea parser particularly? It seems like it was merged before your review changes were applied, so it's worth going back and double checking I caught everything.
> > 
> > 
> > @pipliggins thank you! looks ok - I just added a comment regarding the oxygen_therapy supertype to include all variables that indicate any oxygen therapy was given. I noted from my review the addition of mapping palpitations and edema but these attributes don't seem to be present in the version of the parser I'm looking at - are these attributes still being mapped?
> 
> Thanks Sadie - these two attributes certainly aren't currently being mapped, I'm not actually sure they ever were? I've added those Guinea changes here, it's bit confusing with changes being made to the same parts of the same parser on two different branches!

that's ok, the oxygen_therapy additions look ok, I guess pending any FiO2. I'm not sure where those attributes came from, but will gladly ignore!

**Comment by sadiekelly** [comment]
> Leaving this open until the difference between fio2 and o2_flow is resolved (should O2_flow be mapped as fio2_Lmin?) and we've decided whether fio2 should be included in the oxygen_therapy super type.

@pipliggins I've spoken with Laura and Felipe about these queries and have some answers, as below;

Regarding including FiO2 in the oxygen_therapy supertype, this can be done as any indication of supplemental oxygen in a boolean question can be considered as oxygen therapy. If we don't have the boolean question and just a value for FiO2 in either % or fraction of 1, then it should only be mapped to oxygen therapy where the value is greater than 0.21 or 21 % (these entries would be considered data entry errors as they indicate oxygen as at room air rather than supplemental treatment).

FiO2 captured in L/min is equivalent to other variables capturing oxygen flow volume in L/min so these can be combined into one attribute such as oxygen_flow_volume. FiO2 in L/min is not really an FiO2 measurement but can be converted, the conversion depending on the method of oxygen therapy used. So I think best to label these as oxygen flow volume rather than FiO2. We may need to consider adding a context note where we have different variables capturing oxygen flow volume relating to different oxygen therapy methods e.g. invasive, noninvasive as you picked up in the french database

**Comment by pipliggins** [comment]
@sadiekelly with the 2 new definitions here (oxygen_flow_rate and fio2) - do you think the presence of either/both should be used as an indicator of oxygen therapy in the visit table? Or just one?

**Comment by pipliggins** [comment]
Hi @sadiekelly, would you mind doing a final check through this? I think I've done everything we've discussed, but a second pair of eyes before I merge would be great!

File: isaric/parsers/ccp-ghana.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-ghana.toml:913]: the overall cmtype and cmdose variables are both checkboxes, so if there is more than one option selected for each then it is not possible to associate the type with the dose given of that type. I think it's useful to associate type and L/min as this enables calculation of FiO2, but I think we can only do this where one dose and one type is indicated

Comment (sadiekelly) [reply isaric/parsers/ccp-ghana.toml:913]: daily variables are single select dropdown so they can be associated

File: isaric/parsers/cocoa_pediatrics.toml
Comment (sadiekelly) [inline isaric/parsers/cocoa_pediatrics.toml:570]: can we add context nasal prongs or mask from xygn_thrpintfc_trt_v3?

File: isaric/parsers/ccp-netherlands.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-netherlands.toml:314]: the variable signssymp_L_oxygen also gives L oxygen, I think at admission although not super clear! This can also  be mapped to the oxygen flow attribute with description 'amount of oxygen therapy (L)'

File: isaric/parsers/ccp-guinea.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-guinea.toml:668]: daily_oxygen_cmdose_1 can be added to the oxygen_therapy attribute

File: isaric/parsers/ccp-guinea.toml
Comment (sadiekelly) [inline isaric/parsers/ccp-guinea.toml:1419]: add daily_oxygen_cmdose_1

File: isaric/parsers/isaric-core.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-core.toml:899]: oxygen_cmdose can be added to the oxygen.therapy fields. Probably already covered by oxygen_cmoccur as this is the associated dose, but just in case

File: isaric/parsers/isaric-core.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-core.toml:2011]: oxygen_cmdose can be added to oxygen_flow_rate. This is the maximum flow volume.

File: isaric/parsers/isaric-ecmo.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-ecmo.toml:1211]: i think it might be best to drop the context here, as this data doesn't appear to have been collected in direct relation to the fio2 values, so we don't know whether the fio2 value given relates to one or any of the oxygen therapy types

File: isaric/parsers/isaric-ccpuk.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-ccpuk.toml:1112]: i think it might be best to drop the context values here as I don't think the variables are explicitly related and we don't know whether the fio2 value is related to one of these interventions

File: isaric/parsers/isaric-rapid.toml
Comment (sadiekelly) [inline isaric/parsers/isaric-rapid.toml:1856]: I think best to drop context, as above

File: isaric/parsers/ncov-france.toml
Comment (sadiekelly) [inline isaric/parsers/ncov-france.toml:478]: I'm not sure inspiratory trigger can be used for oxygen flow, seems like a setting on the NIV equipment

**Comment by sadiekelly** [comment]
> Hi @sadiekelly, would you mind doing a final check through this? I think I've done everything we've discussed, but a second pair of eyes before I merge would be great!

hi @pipliggins! I've added a few comments on the updated files - let me know if something is not clear - thanks!

File: isaric/parsers/isaric-ecmo.toml
Comment (pipliggins) [reply isaric/parsers/isaric-ecmo.toml:1211]: Do you mean the whole context section, or just the daily_ecmo_prtrt section?

Comment (sadiekelly) [reply isaric/parsers/isaric-ecmo.toml]: @pipliggins I think the whole context section should be dropped as no specific device was indicated alongside the FiO2 values

**Pull request merged**
